### PR TITLE
MatrixFree: disable overlap of communication/computation for FE_Nothing

### PR DESCRIPTION
**Pull request by peterrum**
as discussed in https://github.com/dealii/dealii/pull/14342#issuecomment-1343349014

### CODE REVIEW

**Comment by sebproell** [comment]
@peterrum Naïve question: is this going to hurt performance? I assume there is a reason to overlap communication and computation, so why is it acceptable to silently disable this option if one uses FE_Nothing? Thanks!

**Comment by peterrum** [comment]
> Naïve question: is this going to hurt performance? I assume there is a reason to overlap communication and computation, so why is it acceptable to silently disable this option if one uses FE_Nothing? Thanks!

In the theory, it should be somewhat slower. In the practice, there should be hardly and difference (it might be even faster): the MPI routines and the user code share the same hardware so that they might deteriorate the performance of each other...

**Comment by kronbichler** [comment]
> the MPI routines and the user code share the same hardware so that they might deteriorate the performance of each other...

This depends on several things, for example whether the MPI library schedules additional threads or whether there are hyperthreads that are able to do the communication. With shared memory, one could try to do things in a more clever way that what we do, e.g. let all but on core compute and the last one communicate. Finally, there is also the question of prefetching and regularity of data access, which gets worse as you deviate from the regular order through cells (and overlapping communication + computation does reschedule computations). We tried to improve the locality in #10012, so the effect should be rather small.

File: include/deal.II/matrix_free/matrix_free.templates.h [hunk 1]
```diff
@@ -1891,6 +1891,19 @@ MatrixFree<dim, Number, VectorizedArrayType>::initialize_indices(
...
+    additional_data.overlap_communication_computation;
+
+  if (overlap_communication_computation)
+    {
+      for (unsigned int no = 0; no < dof_handlers.size(); ++no)
+        for (unsigned int fe_no = 0;
+             fe_no < dof_handlers[no]->get_fe_collection().size();
+             ++fe_no)
+          if (dof_handlers[no]->get_fe(fe_no).n_dofs_per_cell() == 0)
+            overlap_communication_computation = false;
```
Comment (kronbichler) [inline include/deal.II/matrix_free/matrix_free.templates.h:1904]: A comment would be helpful.

File: include/deal.II/matrix_free/matrix_free.templates.h [hunk 2]
```diff
@@ -1906,7 +1919,7 @@ MatrixFree<dim, Number, VectorizedArrayType>::initialize_indices(
     additional_data.cell_vectorization_categories_strict,
     do_face_integrals,
     additional_data.mapping_update_flags_inner_faces != update_default,
-    additional_data.overlap_communication_computation,
+    overlap_communication_computation,
```
Comment (kronbichler) [inline include/deal.II/matrix_free/matrix_free.templates.h:1922]: It looks like this change is now done unconditionally when we detect `FE_Nothing`, not just when we have face integrals. I thought that the problem was only in the latter case?

**Review by kronbichler — APPROVED** [review]
Does this really cover the initial problem in all cases, or is it just a one-off fix? I am referring to the statement in #14342,
> My idea went in a different direction: We explicitly need to look into the indices that each cell handles and check if those depend on other MPI ranks, and then align the face integrals to the indices of the two adjacent cells, and decide whether they go into the MPI-part (index 1) or one of the other two parts (index 0 or 2). The last step is then to make sure that we adhere to the desired dependency between cells and faces (cell integrals always touch the DoFs in a DG sense first).

I am not opposed to take this initial fix, but it feels that we should not forget about the underlying mechanism and look for possible fixes there.

**Comment by kronbichler** [comment]
This is on the list for the 9.6 release. Is this still active, and should it be tackled?

**Comment by peterrum** [comment]
@kronbichler I have rebased this PR and added some comments.

**Review by kronbichler — APPROVED** [review]
Yes, this helps. Thank you!

**Comment by kronbichler** [comment]
You need to reindent:
```diff

Run ./contrib/utilities/check_indentation.sh
diff --git a/tests/matrix_free/fe_nothing_02.cc b/tests/matrix_free/fe_nothing_02.cc
index 72306253b..1d2b8e273 100644
--- a/tests/matrix_free/fe_nothing_02.cc
+++ b/tests/matrix_free/fe_nothing_02.cc
@@ -125,7 +125,7 @@ test(const unsigned int n_refinements)
       VectorTools::interpolate(dof_handler, Fu<dim>(), src);
 
       const auto cell_operation = [&](const auto &matrix_free,
-                                      auto &      dst,
+                                      auto       &dst,
                                       const auto &src,
                                       const auto  range) {
         if (matrix_free.get_cell_range_category(range) != 0)
@@ -145,7 +145,7 @@ test(const unsigned int n_refinements)
       };
 
       const auto face_operation = [&](const auto &matrix_free,
-                                      auto &      dst,
+                                      auto       &dst,
                                       const auto &src,
                                       const auto  range) {
         const auto category = matrix_free.get_face_range_category(range);
```

**Pull request merged**